From 457fa9613ebf57d183c4cca5f5a25ec278835bf8 Mon Sep 17 00:00:00 2001
From: yutingfang <yutingfang@google.com>
Date: Fri, 17 Apr 2026 14:19:19 -0700
Subject: [PATCH 46/49] Fix a regression in ECM mode setting after package
 install

A 3P package installer can set package source to PACKAGE_SOURCE_STORE to
bypass ECM due to lazy evaluation of ECM state. Revert a previous change
to re-instate ECM mode right after package intallation.

Bug: 349318333
Test: atest EnhancedConfirmationManagerTest
Flag: EXEMPT CVE_FIX
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:c06b539f3109383224362ce9ab567a4c78c383e1
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:00822bb857ac91e09b10169a9babc075e94a1d3a
Merged-In: Ifffaf8a004d5c7d20ca60c00c487bc4e54899976
Change-Id: Ifffaf8a004d5c7d20ca60c00c487bc4e54899976
---
 .../server/pm/InstallPackageHelper.java       | 33 ++++++++-----------
 1 file changed, 13 insertions(+), 20 deletions(-)

diff --git a/services/core/java/com/android/server/pm/InstallPackageHelper.java b/services/core/java/com/android/server/pm/InstallPackageHelper.java
index d55ebfa09033..cb97d4c6b3b3 100644
--- a/services/core/java/com/android/server/pm/InstallPackageHelper.java
+++ b/services/core/java/com/android/server/pm/InstallPackageHelper.java
@@ -3320,26 +3320,19 @@ final class InstallPackageHelper {
                 mPm.notifyPackageChanged(packageName, request.getAppId());
             }
 
-            // Set the OP_ACCESS_RESTRICTED_SETTINGS op, which is used by ECM (see {@link
-            // EnhancedConfirmationManager}) as a persistent state denoting whether an app is
-            // currently guarded by ECM, not guarded by ECM or (in Android V+) that this should
-            // be decided later. In Android B, the op's default mode was updated to the
-            // "should be decided later" case, and so this step is now unnecessary.
-            if (!android.permission.flags.Flags.enhancedConfirmationModeApisEnabled()
-                    || !android.security.Flags.extendEcmToAllSettings()) {
-                // Apply restricted settings on potentially dangerous packages. Needs to happen
-                // after appOpsManager is notified of the new package
-                if (request.getPackageSource() == PackageInstaller.PACKAGE_SOURCE_LOCAL_FILE
-                        || request.getPackageSource()
-                        == PackageInstaller.PACKAGE_SOURCE_DOWNLOADED_FILE) {
-                    final int appId = request.getAppId();
-                    mPm.mHandler.post(() -> {
-                        for (int userId : firstUserIds) {
-                            // MODE_ERRORED means that the app is explicitly guarded
-                            setAccessRestrictedSettingsMode(packageName, appId, userId,
-                                    AppOpsManager.MODE_ERRORED);
-                        }
-                    });
+            // Apply restricted settings on potentially dangerous packages. Needs to happen
+            // after appOpsManager is notified of the new package
+            if (request.getPackageSource() == PackageInstaller.PACKAGE_SOURCE_LOCAL_FILE
+                    || request.getPackageSource()
+                    == PackageInstaller.PACKAGE_SOURCE_DOWNLOADED_FILE) {
+                final int appId = request.getAppId();
+                Slog.i(TAG, "PackageSource is local file or downloaded file");
+                for (int userId : firstUserIds) {
+                    // MODE_ERRORED means that the app is explicitly guarded
+                    Slog.i(TAG, "Set OP_ACCESS_RESTRICTED_SETTINGS to MODE_ERRORED for user  "
+                                    + userId);
+                    setAccessRestrictedSettingsMode(packageName, appId, userId,
+                            AppOpsManager.MODE_ERRORED);
                 }
             }
 
-- 
2.55.0

